### PR TITLE
Bootstrap test suite for front-end app

### PR DESCRIPTION
**Pull request by mongrelion**
Things that you will find in this pull request:
- Small refactor of the front-end app: I moved the routes definitions out of the `server.js` file into their own modules to make adding tests more easy. Tests are WIP
- Setup the test suite for unit, functional and end-to-end tests: Chai, Mocha, Phantomjs, Casperjs, Makefile with tasks for running the tests, preparing containers, etc.
- Added some basic unit tests for the API
- Added some basic e2e tests for testing the following flows: home page, login, cart interaction, checkout.

Read the README file for instructions on how to run the tests.

### CODE REVIEW

**Comment by sashkachan** [comment]
Can you squash the commits?

Maybe group by unit / functional and e2e testing. I think it'd be better if they were in separate PR's, but it's ok if just separate commits.

File: sockshop/front-end/Makefile [hunk 1]
```diff
@@ -0,0 +1,37 @@
+.PHONY: test coverage
```
Comment (sashkachan) [inline sockshop/front-end/Makefile:1]: Don't you need more PHONY targets in here?

**Comment by sashkachan** [comment]
What node version should I be using?
Tried with v0.10.46

Get thrown the following exception:

``````
[sashka@Sashka-MBP]$ make test
=============================================================================
Writing coverage object [/Users/sashka/Sites/microservices-demo/sockshop/front-end/coverage/coverage.json]
Writing coverage reports at [/Users/sashka/Sites/microservices-demo/sockshop/front-end/coverage]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 100% ( 2/2 )
Branches     : 100% ( 0/0 )
Functions    : 100% ( 1/1 )
Lines        : 100% ( 2/2 )
================================================================================

/Users/sashka/Sites/microservices-demo/sockshop/front-end/test/e2e/config.js:4
  casper.options.viewportSize = {
  ^
ReferenceError: casper is not defined
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/test/e2e/config.js:4:3
    at Object.<anonymous> (/Users/sashka/Sites/microservices-demo/sockshop/front-end/test/e2e/config.js:8:2)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/test/e2e/cart_test.js:3:3
    at Object.<anonymous> (/Users/sashka/Sites/microservices-demo/sockshop/front-end/test/e2e/cart_test.js:44:2)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/mocha/lib/mocha.js:220:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/mocha/lib/mocha.js:217:14)
    at Mocha.run (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/mocha/lib/mocha.js:485:10)
    at Object.<anonymous> (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/mocha/bin/_mocha:403:18)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at runFn (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/istanbul/lib/command/common/run-with-cover.js:122:16)
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/istanbul/lib/command/common/run-with-cover.js:251:17
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/istanbul/lib/util/file-matcher.js:68:16
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/async/lib/async.js:52:16
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/async/lib/async.js:361:13
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/async/lib/async.js:52:16
    at done (/Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/async/lib/async.js:246:17)
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/async/lib/async.js:44:16
    at /Users/sashka/Sites/microservices-demo/sockshop/front-end/node_modules/async/lib/async.js:358:17
    at LOOP (fs.js:1370:14)
    at process._tickCallback (node.js:458:13)
make: *** [test] Error 8```
``````

File: sockshop/front-end/Makefile (hunk 1 above)
Comment (mongrelion) [inline sockshop/front-end/Makefile:1]: Not really. I tried every single task in the Makefile (I wrote it myself) and the only two tasks that have matching folders are these two.

**Comment by mongrelion** [comment]
@alex-glv The dependencies list is here: https://github.com/microservices-demo/microservices-demo/blob/77ec0ebabe433e7625b8d913d87bca6261b354de/sockshop/front-end/README.md#dependencies
Did you follow the instructions in the README file? you should at some point run the task `make deps` for all the nodejs modules to be installed. If you have and you still see this error, can you `ls node_modules | grep casper` and dump the output here?

**Comment by mongrelion** [comment]
@alex-glv thanks for helping me out with the debugging. I submitted the patch that fixes the test suite.
As we're the project into smaller repos, the squashing doesn't make much sense anymore.
Can I :shipit:?

**Comment by mongrelion** [comment]
Merging as discussed with @alex-glv on Slack.

**Pull request merged**